From 74a9e297ff881de29f55c0c56f98fbfec93299ea Mon Sep 17 00:00:00 2001
From: Xu Liang <liang.xu@intel.com>
Date: Fri, 16 Oct 2020 23:03:13 +0800
Subject: [PATCH] PONRTSYS-8089: GSW API - update algorithm for PIR calculation
 because IBS is fixed and calculated based on CIR. PIR can not be less than
 CIR.

---
 .../net/ethernet/lantiq/switch-api/gsw_flow_core.c | 67 ++++++++++++++++------
 1 file changed, 50 insertions(+), 17 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
index d16904d99c3f..ea35db78b9f1 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_flow_core.c
@@ -2258,6 +2258,25 @@ static int calc_mtoken(u32 mrate, u32 *ibsid, u32 *expont, u32 *mants)
 	return -1;
 }
 
+static int calc_mtoken_pir(u32 mrate, u32 ibsid, u32 *expont, u32 *mants)
+{
+	static const u32 ibs_table[] = {8 * 8, 32 * 8, 64 * 8, 96 * 8};
+	u32 ibs = ibs_table[ibsid];	/* ibsid is 2 bit value generated by
+					 * function calc_mtoken */
+	u32 exp, mant;
+
+	for (exp = 0; exp < 16; exp++) {
+		mant = ((ibs * 25000) >> exp) / mrate;
+
+		if (mant < (1 << 10)) {
+			*expont = exp;
+			*mants = mant;
+			return 0;
+		}
+	}
+	return -1;
+}
+
 static u32 mratecalc_3_1(u32 ibsid, u32 expont, u32 mants)
 {
 	u32 mrate = 0;
@@ -2302,6 +2321,22 @@ static int calc_mtoken_3_1(u32 mrate, u32 *ibsid, u32 *expont, u32 *mants)
 	return 0;
 }
 
+static int calc_mtoken_pir_3_1(u32 mrate, u32 ibs, u32 *expont, u32 *mants)
+{
+	u32 exp, mant;
+
+	for (exp = 0; exp < 16; exp++) {
+		mant = (((ibs * 200000 + exp / 2) >> exp) + mrate / 2) / mrate;
+
+		if (mant > 0 && mant < (1 << 10)) {
+			*expont = exp;
+			*mants = mant - 1;
+			return 0;
+		}
+	}
+	return -1;
+}
+
 #endif /*CONFIG_LTQ_QOS */
 #if defined(CONFIG_LTQ_VLAN) && CONFIG_LTQ_VLAN
 static void reset_vlan_sw_table(void *cdev)
@@ -7297,14 +7332,13 @@ GSW_return_t GSW_QoS_MeterCfgSet(void *cdev,
 
 	mid = parm->nMeterId;
 
-	if (IS_VRSN_30_31(gswdev->gipver)) {
+	if (VRSN_CMP(gswdev->gipver, LTQ_GSWIP_3_0) >= 0) {
 		if (mid >= gswdev->num_of_meters) {
 			ret = GSW_statusErr;
 			goto UNLOCK_AND_RETURN;
 		}
 
-
-		if (IS_VRSN_31(gswdev->gipver)) {
+		if (VRSN_CMP(gswdev->gipver, LTQ_GSWIP_3_1) >= 0) {
 			/*Check whether it is InUSE,if not InUse,return ERROR*/
 			if (!(gswdev->meter_idx[mid].IndexInUse)) {
 				ret = GSW_statusErr;
@@ -7339,7 +7373,7 @@ GSW_return_t GSW_QoS_MeterCfgSet(void *cdev,
 		rate = parm->nRate;
 
 		if (rate) {
-			if (IS_VRSN_31(gswdev->gipver))
+			if (VRSN_CMP(gswdev->gipver, LTQ_GSWIP_3_1) >= 0)
 				ret = calc_mtoken_3_1(rate, &ibs, &exp, &mant);
 			else
 				ret = calc_mtoken(rate, &ibs, &exp, &mant);
@@ -7359,7 +7393,7 @@ GSW_return_t GSW_QoS_MeterCfgSet(void *cdev,
 			GSW_PCE_TCM_CIR_MANT_MANT_SIZE, mant);
 
 		/* Rate Counter iBS */
-		if (IS_VRSN_31(gswdev->gipver))
+		if (VRSN_CMP(gswdev->gipver, LTQ_GSWIP_3_1) >= 0)
 			gsw_w32(cdev, GSW_PCE_TCM_IBS_IBS_OFFSET,
 				GSW_PCE_TCM_IBS_IBS_SHIFT,
 				GSW_PCE_TCM_IBS_IBS_SIZE_GSWIP_3_1, ibs);
@@ -7369,13 +7403,15 @@ GSW_return_t GSW_QoS_MeterCfgSet(void *cdev,
 				GSW_PCE_TCM_IBS_IBS_SIZE_GSWIP_3_1, ibs);
 
 		/* Calc the Rate and convert to MANT and EXP*/
-		rate = parm->nPiRate;
+		rate = max(parm->nRate, parm->nPiRate);
 
 		if (rate) {
-			if (IS_VRSN_31(gswdev->gipver))
-				ret = calc_mtoken_3_1(rate, &ibs, &exp, &mant);
-			else
-				ret = calc_mtoken(rate, &ibs, &exp, &mant);
+			if (VRSN_CMP(gswdev->gipver, LTQ_GSWIP_3_1) >= 0) {
+				ret = calc_mtoken_pir_3_1(rate, ibs, &exp,
+							  &mant);
+			} else {
+				ret = calc_mtoken_pir(rate, ibs, &exp, &mant);
+			}
 			if (ret) {
 				ret = GSW_statusErr;
 				goto UNLOCK_AND_RETURN;
@@ -7390,10 +7426,6 @@ GSW_return_t GSW_QoS_MeterCfgSet(void *cdev,
 		gsw_w32(cdev, GSW_PCE_TCM_PIR_MANT_MANT_OFFSET,
 			GSW_PCE_TCM_PIR_MANT_MANT_SHIFT,
 			GSW_PCE_TCM_PIR_MANT_MANT_SIZE, mant);
-		/* Rate Counter iBS */
-		/*		gsw_w32(cdev, GSW_PCE_TCM_IBS_IBS_OFFSET, */
-		/*			GSW_PCE_TCM_IBS_IBS_SHIFT,*/
-		/*			GSW_PCE_TCM_IBS_IBS_SIZE, ibs);*/
 		gsw_w32(cdev, GSW_PCE_TCM_CTRL_TMOD_OFFSET,
 			GSW_PCE_TCM_CTRL_TMOD_SHIFT,
 			GSW_PCE_TCM_CTRL_TMOD_SIZE, parm->eMtrType);
@@ -7402,10 +7434,12 @@ GSW_return_t GSW_QoS_MeterCfgSet(void *cdev,
 			GSW_PCE_TCM_CTRL_TCMEN_SHIFT,
 			GSW_PCE_TCM_CTRL_TCMEN_SIZE, parm->bEnable);
 
-		if (IS_VRSN_31(gswdev->gipver))
+		if (VRSN_CMP(gswdev->gipver, LTQ_GSWIP_3_1) >= 0) {
 			gsw_w32(cdev, GSW_PCE_TCM_CTRL_BLIND_OFFSET,
 				GSW_PCE_TCM_CTRL_BLIND_SHIFT,
-				GSW_PCE_TCM_CTRL_BLIND_SIZE, parm->nColourBlindMode);
+				GSW_PCE_TCM_CTRL_BLIND_SIZE,
+				parm->nColourBlindMode);
+		}
 	} else {
 		if (mid > 7) {
 			ret = GSW_statusErr;
@@ -28163,4 +28197,3 @@ static void gsw_init_fn_ptrs(struct core_ops *ops)
 	ops->gsw_irq_ops.IRQ_Disable					= GSW_Irq_disable;
 
 }
-
